### ISSUE TITLE
Incorrect nullability warning in local function

### ISSUE THREAD

**Issue by JohanLarsson** [dotnet/roslyn#40157]
```cs
void M()
{
    if (Try(out var s))
    {
        int Local()
        {
            //     ↓ CS8602 here with 16.4
            return s.Length;
        }
    }
}

bool Try([NotNullWhen(true)] out string? text)
{
    text = string.Empty;
    return true;
}
```

**Comment by jnm2** [comment]
I think this is https://github.com/dotnet/roslyn/issues/38548?

**Comment by JohanLarsson** [comment]
Looks similar, I would add tests for both.

**Comment by jcouv** [comment]
This is a problem with analysis of local functions, which doesn't seem particularly related to the attribute.
Here's a [repro without attribute](https://sharplab.io/#v2:D4AQDABCCMB0AiBLAhgcwHYHsDOAXRAxtrAMKYAmApgILrIA2AntotgNwCwAUNwMToBXevWQAjepQiU64ytxAAmCCW4BveQBYIAWQAUMMAH4I2AJRruEKxEvXEAMwi7sEVhEHDzXaxHXfrAL62VpT02HL+Vn4+PojouBAAMpgEDLpeMdbRmTEA9LkxgMmEygDKABwAbGBKABaUAE6SAO6IuDUQ0BWwGsE5ViAA7CawidKobZyRMUFTMzPcopiY9BAAKvWMugDaAHKYuDtC9ADqdei6uPUClKYAuhCYAgkGxriUAB64XtlWb58QAF4oNAwLAAKIAWwADrhGJMfIMIJdrpM5kA):
``` csharp
using System.Diagnostics.CodeAnalysis;

#nullable enable
class C
{
void M(string? s)
{
    
    if (s is null)
    {
    }
    else
    {
        int Local()
        {
            //     ↓ CS8602 here with 16.4
            return s.Length;
        }
    }
}

bool Try([NotNullWhen(true)] out string? text)
{
    text = string.Empty;
    return true;
}
}
```

**Comment by RikkiGibson** [comment]
When a local function is never used, it is analyzed using variable state from the declared type of the variables. I think we should consider instead handling the case where the local function is never used by analyzing it with the state at the point of its declaration. When the code is changed to contain one or more uses of the local function, we can use our existing strategy of `Join`ing the states at the usage points into the top state. That is to say, we would no longer care where the local function is declared when determining variable state within the local function.

**Comment by RikkiGibson** [comment]
There is at least one other issue with the same root cause as this which I'd like to dupe against this.

**Comment by jnm2** [comment]
@RikkiGibson [In Gitter](https://gitter.im/dotnet/csharplang?at=5f00a9bcfa0c9221fc7b148e), @mikernet made the point which I agree with that local functions should not warn about nullability of variables in the containing method until they begin to be called.

**Comment by RikkiGibson** [comment]
It's not obvious to me what the best heuristic is here. Consider something like:

```cs
class C
{
    string? Prop { get; }

    void M1()
    {
        var local1 = "b";
        string? local2 = null;

        return;

        void localFunc()
        {
            local1.ToString(); // no warning here is good.
            local2.ToString(); // no warning here seems bad?
            Prop.ToString(); // no warning here seems bad.
        }
    }
}
```

But maybe the user was intending to do something like this when they finally called the local function:
```cs
class C
{
    string? Prop { get; }

    void M2()
    {
        var local1 = "b";
        string? local2 = null;

        if (Prop is not null && local2 is not null)
        {
            localFunc();
        }

        return;

        void localFunc()
        {
            // no warnings should happen at all this time.
            local1.ToString();
            local2.ToString();
            Prop.ToString();
        }
    }
}
```

It's a bit late to be making the change, but it might still be possible to say: `var` variables get their nullable annotation from their initializer, as they did earlier on in the feature development, **but**, we simply don't give the CS8600 warning when null is assigned to them.

I think this might put us in the best place to analyze local functions that haven't been called yet. And I do think this is an important scenario. People often implement a local function before calling it and it's a pain to only get the nullable warnings once the implementation is "done".

Also, we've received reams of issue reports where people are confused that they assigned a non-nullable thing to a `var` and it ended up having a nullable type when they do Quick Info on the variable.

**Comment by jjonescz** [comment]
Fixed by https://github.com/dotnet/roslyn/pull/71679.
